cpp/src/array_utils.h

```diff
@@ -309,3 +309,20 @@ std::string toFloatVectorString(std::vector<data_t> vec) {
   return toFloatVectorString<dist_t, data_t, scalefactor>(vec.data(),
                                                           vec.size());
 }
+
+/**
+ * Convert a 2D vector of float to NDArray<float, 2>
+ */
+NDArray<float, 2> vectorsToNDArray(std::vector<std::vector<float>> vectors) {
+  int numVectors = vectors.size();
+  int dimensions = numVectors > 0 ? vectors[0].size() : 0;
+  std::array<int, 2> shape = {numVectors, dimensions};
+
+  // flatten the 2d array into the NDArray's underlying 1D vector
```

Comment:
Rather than iterating over the outer and inner vectors, we should try to utilize the [underlying data access that `std::vector` provides](https://en.cppreference.com/w/cpp/container/vector/data).  Be careful with this though as there may be caveats around vector memory allocation boundaries and actual vector item counts

Reply:
I am still iterating over each vector because I added a check to validate that each vector size is identical. But now I am preallocating the space I need in the flattened array to avoid resizing, and using `std::memcpy` to do the copying